File: app/src/main/java/com/skyline/app/DestinationAdapter.java
package com.skyline.app;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.skyline.app.databinding.ItemDestinationBinding;
import com.skyline.model.Destination;
import java.util.List;

public class DestinationAdapter extends RecyclerView.Adapter<DestinationAdapter.DestinationViewHolder> {
    private List<Destination> items;
    private OnItemClickListener listener;

    public interface OnItemClickListener {
        void onItemClick(Destination item);
    }

    public DestinationAdapter(List<Destination> items, OnItemClickListener listener) {
        this.items = items;
        this.listener = listener;
    }

    @NonNull
    @Override
    public DestinationViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemDestinationBinding binding = ItemDestinationBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new DestinationViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull DestinationViewHolder holder, int position) {
        holder.bind(items.get(position));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    class DestinationViewHolder extends RecyclerView.ViewHolder {
        private ItemDestinationBinding binding;

        public DestinationViewHolder(ItemDestinationBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(Destination item) {
            String url = item.getImageUrl();
            if (url != null && !url.isEmpty()) {
                Glide.with(binding.imgDestination.getContext())
                        .load(url)
                        .placeholder(R.drawable.img_destination_hochiminh)
                        .error(R.drawable.img_destination_hochiminh)
                        .into(binding.imgDestination);
            } else if (item.getImageRes() != 0) {
                binding.imgDestination.setImageResource(item.getImageRes());
            } else {
                binding.imgDestination.setImageResource(R.drawable.img_destination_hochiminh);
            }
            binding.tvCountry.setText(item.getCountry());
            binding.tvDestinationTitle.setText(item.getTitle());
            binding.getRoot().setOnClickListener(v -> listener.onItemClick(item));
        }
    }
}

File: app/src/main/java/com/skyline/app/DestinationAdapter.kt
package com.skyline.app

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.skyline.app.databinding.ItemDestinationBinding
import com.skyline.model.Destination

class DestinationAdapter(
    private val items: List<Destination>,
    private val onClick: (Destination) -> Unit = {}
) : RecyclerView.Adapter<DestinationAdapter.DestinationViewHolder>() {

    inner class DestinationViewHolder(
        private val binding: ItemDestinationBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(item: Destination) = with(binding) {
            imgDestination.setImageResource(item.imageRes)
            tvCountry.text = item.country
            tvDestinationTitle.text = item.title
            root.setOnClickListener { onClick(item) }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DestinationViewHolder {
        val binding = ItemDestinationBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return DestinationViewHolder(binding)
    }

    override fun onBindViewHolder(holder: DestinationViewHolder, position: Int) {
        holder.bind(items[position])
    }

    override fun getItemCount(): Int = items.size
}